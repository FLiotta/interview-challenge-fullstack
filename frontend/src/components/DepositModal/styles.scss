.depositmodal {

  &__footer {
    display: flex;
    justify-content: flex-end;
    margin-top: 35px;
    & > button {
      margin-left: 15px;
    }
  }

  &__payment {
    line-height: 1.5;
    font-size: 0.6rem;
    margin-bottom: 0px;
    margin-top: 15px;
    margin-bottom: 25px;
  }
}