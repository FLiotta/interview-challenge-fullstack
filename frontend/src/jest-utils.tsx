// @Packages
import { ReactNode } from 'react';
import { render as rtlRender } from '@testing-library/react';
import { Provider } from 'react-redux';

// @Project
import store from 'store';
import { JsxElement } from 'typescript';

function render(
  ui: ReactNode | JsxElement | any,
  renderOptions: any = { }
) {
  function Wrapper({ children }: any) {
    return (
      <Provider store={store}>{children}</Provider>
    )
  }
  return rtlRender(ui, {
     wrapper: Wrapper, 
     ...renderOptions
    })
}

export * from '@testing-library/react'

export { render }
