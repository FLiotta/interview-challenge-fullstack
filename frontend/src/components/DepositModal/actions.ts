// @Project
import { Account } from 'interfaces';

export const DEPOSITMODAL_OPEN = '[DEPOSIT MODAL] OPEN';
export const DEPOSITMODAL_CLOSE = '[DEPOSIT MODAL] CLOSE';

export const openDepositModal = (acc: Account) => ({
  type: DEPOSITMODAL_OPEN,
  payload: { acc }
})

export const closeDepositModal = () => ({
  type: DEPOSITMODAL_CLOSE
})