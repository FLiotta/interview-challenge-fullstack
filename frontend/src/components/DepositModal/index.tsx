// @Packages
import React, { useState } from 'react';
import Modal, { Styles } from 'react-modal';
import { useForm } from 'react-hook-form';
import { useSelector, useDispatch } from 'react-redux';
import cogoToast from 'cogo-toast';

// @Project
import AccountService from 'services/account';
import { Account } from 'interfaces';

// @Own
import './styles.scss';

interface FormPayload { amount: number }

const customStyles: Styles = {
  content: {
    top: '50%',
    left: '50%',
    right: 'auto',
    bottom: 'auto',
    marginRight: '-50%',
    transform: 'translate(-50%, -50%)',
    maxWidth: '100%',
    width: 550
  },
  overlay: {
    backgroundColor: 'rgba(0,0,0,0.3)',
    zIndex: 15
  }
};

Modal.setAppElement('#root');

interface IProps {
  visible: boolean,
  onClose: () => void
  onSuccess: () => void
  account: Account
}

const DepositModal: React.FC<IProps> = ({
  visible,
  onClose,
  onSuccess,
  account
}) => {
  const { register, handleSubmit, setValue } = useForm();
  const [loading, setLoading] = useState<boolean>(false);

  const handleFormSubmit = (payload: FormPayload) => {
    if(account) {
      const { amount } = payload;

      setLoading(true);
      console.log(account);
      AccountService.deposit(account.id, Number(amount))
        .then((resp) => {
          console.warn("Se esta simulando un tiempo de 2s solo por fines esteticos =)")
          cogoToast.success('Deposito acreditado con exito', {
            position: 'bottom-right'
          })
          console.log(resp);
          onSuccess();
          setValue('amount', undefined);
          setLoading(false)
        })
        .catch(() => {
          cogoToast.error('Hubo un problema con el procesamiento de tu deposito.', {
            position: 'bottom-right'
          })
          setLoading(false)
        });
    }
  }


  return (
    <Modal
      isOpen={visible}
      onRequestClose={onClose}
      style={customStyles}
      contentLabel="Modal de depositos"
    >
      <h4>ðŸ’° Depositos</h4>
      <p>
        Estas proximo a hacer un deposito en tu cuenta #{account?.id} ({account?.name}).
        <br/><br/>
        Ingresa acontinuacion el valor deseado:
      </p>
      <form 
        id="depositModalForm"
        onSubmit={handleSubmit(handleFormSubmit)}
      >
        <div className="form-group">
          <input
            {...register('amount', { required: true })}
            placeholder="e.g: 1500"
            className="form-control"
          />
        </div>
      </form>

      <p className="depositmodal__payment">Se usara el metodo de pago configurado en tu cuenta por defecto para procesar la operacion.</p>

      <div className="depositmodal__footer">
        <button 
          type="button" 
          className="btn btn-brand-secondary" onClick={onClose}
          disabled={loading}
        >
          Cancelar
        </button>
        <button 
          type="submit" 
          className="btn btn-brand"
          disabled={loading}
          form="depositModalForm"
        >
          {loading
            ? <>
              <span className="spinner-border spinner-border-sm me-2" role="status" aria-hidden="true"></span>
              Procesando...
            </>
            : 'Depositar'
          }
        </button>
      </div>
    </Modal>
  )
}

export default DepositModal;