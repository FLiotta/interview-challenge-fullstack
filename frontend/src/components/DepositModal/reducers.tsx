// @Packages
import { AnyAction } from "redux";

// @Project
import { Account } from 'interfaces';

// @Own
import { DEPOSITMODAL_OPEN, DEPOSITMODAL_CLOSE } from './actions';

export interface IState {
  visible: boolean
  account?: Account
}

const defaultState: IState = {
  visible: false,
  account: undefined
}

export default (state = defaultState, action: AnyAction) => {
  switch(action.type) {
    case DEPOSITMODAL_OPEN:
      return {
        visible: true,
        account: action.payload.acc
      }
    case DEPOSITMODAL_CLOSE:
      return {
        visible: false,
        account: {}
      }
    default: 
      return state;
  }
}