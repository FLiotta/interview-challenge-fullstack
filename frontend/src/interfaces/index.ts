// @Packages
import { ThunkDispatch } from "redux-thunk";

// @Project
import { IStoreState } from 'reducers/rootReducer';
import { AnyAction } from "redux";

export interface IThunkDispatch extends ThunkDispatch<IStoreState, any, AnyAction> {}

export interface BackendResponse<T> {
  error?: string
  message?: string
  data: T | T[] | any
  metadata?: {
    total: number
  }
}

export interface Profile {
  id: number
  username: string,
  first_name?: string
  last_name?: string
  email?: string
  is_superuser: boolean,
  is_staff: boolean
}

export interface Currency {
  id: number,
  name: string,
  symbol: string
}

export interface Account {
  id: number,
  name?: string,
  currency?: Currency
  currency_id?: number,
  funds: number,
  owner_id?: number,
  owner?: any
}

export interface Operation {  
  id: number
  amount: number
  currency: Currency
  operation_type: 'transfer' | 'deposit'
  receiver: any
  sender: any
  date: string | Date
}