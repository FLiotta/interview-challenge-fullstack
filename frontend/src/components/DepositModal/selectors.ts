import { Account } from 'interfaces';
import { IStoreState } from 'reducers/rootReducer';

export const selectSelectedAccount = (state: IStoreState): Account | undefined => state.depositmodal.account;
export const selectVisible = (state: IStoreState): boolean => state.depositmodal.visible;