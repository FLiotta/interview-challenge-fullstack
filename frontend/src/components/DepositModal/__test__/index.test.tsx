// @Packages
import { getByTestId } from '@testing-library/react';
import { render, fireEvent } from 'jest-utils';
import ReactModal from 'react-modal';

// @Project
import { Account } from 'interfaces';
import DepositModal from '../index';
import ReactDOM from 'react-dom';

const MOCKACCOUNT: Account = {
  id: 1,
  funds: 0,
  deposit_address: "KQWEMRKFWEQRMKWQER"
}

describe('DepositModal component', () => {

  it('Should match snapshot', () => {
    const { container } = render(
      <DepositModal
        visible
        onClose={() => {}}
        onSuccess={() => {}}
        account={MOCKACCOUNT}
      />,
      { 
        container: document.body
      }
    )

    expect(container).toMatchSnapshot();
  });

  it('Fires onClose on close button click', () => {
    const mockedFn = jest.fn();
    const closeBtnTestId = 'depositmodal-closebtn-testid'
    
    const { container } = render(
      <DepositModal
        visible
        onClose={mockedFn}
        onSuccess={() => {}}
        account={MOCKACCOUNT}
      />, {
        container: document.body,
      }
    );

    const closeBtn = getByTestId(container, closeBtnTestId);

    fireEvent.click(closeBtn);

    expect(mockedFn).toHaveBeenCalledTimes(1);
  });
})