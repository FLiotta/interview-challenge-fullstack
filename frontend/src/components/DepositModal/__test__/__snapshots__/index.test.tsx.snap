// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DepositModal component Should match snapshot 1`] = `
<body
  class="ReactModal__Body--open"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div
    class="ReactModalPortal"
  >
    <div
      class="ReactModal__Overlay"
      style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(0, 0, 0, 0.3); z-index: 15;"
    >
      <div
        aria-label="Modal de depositos"
        aria-modal="true"
        class="ReactModal__Content"
        role="dialog"
        style="position: absolute; top: 50%; left: 50%; border: 1px solid #ccc; background: rgb(255, 255, 255); overflow: auto; border-radius: 4px; outline: none; padding: 20px; margin-right: -50%; transform: translate(-50%, -50%); max-width: 100%; width: 550px;"
        tabindex="-1"
      >
        <h4>
          ðŸ’° Depositos
        </h4>
        <p>
          Estas proximo a hacer un deposito en tu cuenta #
          1
           (
          ).
          <br />
          <br />
          Ingresa acontinuacion el valor deseado:
        </p>
        <form
          id="depositModalForm"
        >
          <div
            class="form-group"
          >
            <input
              class="form-control"
              name="amount"
              placeholder="e.g: 1500"
            />
          </div>
        </form>
        <p
          class="depositmodal__payment"
        >
          Se usara el metodo de pago configurado en tu cuenta por defecto para procesar la operacion.
        </p>
        <div
          class="depositmodal__footer"
        >
          <button
            class="btn btn-brand-secondary"
            data-testid="depositmodal-closebtn-testid"
            type="button"
          >
            Cancelar
          </button>
          <button
            class="btn btn-brand"
            data-testid="depositmodal-submitbtn-testid"
            form="depositModalForm"
            type="submit"
          >
            Depositar
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
